### PR TITLE
Fixes #26 pypandoc filter added which convert yml to yaml

### LINKED ISSUE #26: Improper conversion from md to rst
The automated conversion from md to rst conversion sometimes leads to the generated html differ from what would have been generated by original md file itself. 

Using Yaydoc's own [README](https://github.com/fossasia/yaydoc/blob/4e632048c6dd35d011aa3d9d882a1825b5b37e33/README.md)

converting it to rst, the resultant page look like this
md2rst

Ideally It should have looked like this
md2html

I made a set of minimal changes to the README such that when the resultant file is converted to rst and rendered as html, we get what we expect.

The diff of the two files can be found [here](https://www.diffchecker.com/eaezT6de)

Similar process was repeated for other README files.

## Findings

In order to convert to rst properly, the md file should have the following properties. This list is not exhaustive.

- All headings must have a blank line before them.
- Code blocks should begin from beginning of line
- code block to render `yml` files should be named `yaml` not `yml`
- A list must be preceded by a blank line

## Proposed Solution

One idea I could think of would be to identify as many of these fixes as we can and apply all of them to the markdown before passing it to pandoc/sphinx.

### PR DESCRIPTION
**Pull request by sch00lb0y**
I wrote a simple parser for changing yml to yaml in code block for proper md to rst conversion by pypandoc. @niranjan94 @pri22296 @imujjwal96 please check this

### CODE REVIEW

File: scripts/parser.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                            splitted_syntax_highligter[0] = '```'
+                            words = ''.join([x for x in splitted_syntax_highligter])
+                            token.append((words, 'SYNTAX HIGHLIGHTER'))
+                        else:
+                            token.append((words, 'WORD'))
+                        words = ''
+                    token.append((text[pos], 'SPACE'))
+                    pos = pos + 1
+                elif text[pos] == '\n':
+                    if len(words) > 0:
```
Comment (pri22296) [inline scripts/parser.py]: Move this block to a seperate function. It's being repeated twice

**Comment by pri22296** [comment]
@sch00lb0y Instead of doing this, why not just replace every instance of \`\`\`yml with \`\`\`yaml ?

**Comment by sch00lb0y** [comment]
@pri22296 Yeah we can simply replace ```yml with ```yaml. but in future if we ant add extra support for conversion. we just have to improve the parser. not only it replace yml with ymal it also add new line to code block

**Comment by pri22296** [comment]
That's a good idea. Could you maybe make the code more modular, so that adding other conversions would not require modification of the parser. basically, could you seperate the core parser, and the conversions it performs.

**Comment by sch00lb0y** [comment]
@pri22296 i have changed as you suggested. even i have changed the parser from simple loop to abstract syntax tree. please test this

**Comment by pri22296** [comment]
parser method is raising error
```
    return nodes[index].text + parser(nodes, index + 1)
TypeError: Can't convert 'NoneType' object to str implicitly
```
lexer method seems to be working fine.

**Comment by sch00lb0y** [comment]
i have tested the algo with open event server [readme](https://raw.githubusercontent.com/fossasia/open-event-orga-server/development/README.md). it working fine,but when i try to parse big markdown file i got this error
`RecursionError: maximum recursion depth exceeded in comparison` so i have to add the recursion limit like this 
`sys.setrecursionlimit(35000)`. I'm not sure how big file can be parsed by python. I think this is the right time for choosing the base language that we going to use in this project. from my side i suggest we can go with `go-lang`. @niranjan94 @pri22296 @imujjwal96  @yukiisbored @jig08 please give me your input.
Thanks

**Comment by imujjwal96** [comment]
@sch00lb0y could you please create a new issue so that the discussion about the base language can be done there?

**Comment by pri22296** [comment]
I would suggest that instead of creating yet another markdown parser, we use an existing markdown parser. It would be more efficient.

**Comment by sch00lb0y** [comment]
yeah that's a good idea we can fork that repo and make changes but i don't think so it'll be easy for us to modify recommonmark.

**Comment by pri22296** [comment]
Instead of forking, we can just import the package and extend it. It would still be easier than creating our own parser. Let's not reinvent the wheel.

**Comment by sch00lb0y** [comment]
extending? i'm not getting. could you elaborate. here we not making our own recommonmark or pypandoc. we just adding extra parsing that needed by pypandoc for proper conversion. it'll be hardly around 4 or 5 support. we just add that alone

**Comment by pri22296** [comment]
What i'm trying to say that let's use a pre existing lexer, and exend it's methods to perform the required fixes.  take a look at the [output](https://paste.ee/p/rei4s) of `mistune.BlockLexer.parse` on yaydoc's readme.
All i'm trying to say is that let's not reimplement a markdown lexer.

**Comment by sch00lb0y** [comment]
that lexer having several token for example they have nodes for list. if we use that we have to handle each and every token. it'll be tedious task for us to handle. in our lexer we creating node based on our need

**Comment by pri22296** [comment]
Oh I see. how about using Pandoc filters? 

from the pandoc docs
```
Pandoc consists of a set of readers and writers. When converting a document from one
format to another, text is parsed by a reader into pandoc’s intermediate representation
of the document—an “abstract syntax tree” or AST—which is then converted by the
writer into the target format. The pandoc AST format is defined in the module
Text.Pandoc.Definition in pandoc-types.

A “filter” is a program that modifies the AST, between the reader and the writer:

INPUT --reader--> AST --filter--> AST --writer--> OUTPUT
```

Creating filters can be done using https://github.com/jgm/pandocfilters

**Comment by sch00lb0y** [comment]
i have changed our own implementation of AST to pandoc filter as @pri22296 suggested. @imujjwal96 @pri22296 please review this.
Thanks!!

File: scripts/yml_filter.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env python
+from pandocfilters import CodeBlock, toJSONFilter
+
+
+def yml_to_yaml(key, val, fmt, meta):
+    if key == 'CodeBlock':
+    	[[ident, classes, keyvals], code] = val
+    	if len(classes) > 0:
+    		if classes[0] == u'yml':
```
Comment (pri22296) [inline scripts/yml_filter.py]: use 4 space indentation

**Comment by sch00lb0y** [comment]
@pri22296 I have checked the code style using pycodestyle. it's fine. i think we should integrate some linter in github to avoid manual review for this

**Comment by sch00lb0y** [comment]
@pri22296 difference [link](https://www.diffchecker.com/YpQYyeQS)

**Pull request merged**